### PR TITLE
Fix 3.0->3.1 regression in JSON serializing nested concurrent dictionaries

### PR DESCRIPTION
**Pull request by layomia**
## Description

Fixes https://github.com/dotnet/runtime/issues/670, where there was a regression between 3.0 and 3.1 where serializing nested concurrent dictionaries went from supported to unsupported. The issue also exists in master/5.0. https://github.com/dotnet/runtime/pull/784 fixes it in master, and this PR ports the fix to 3.0.

In addition to the fix described above, this PR also also has a deserialization fix for an `InvalidCastException` thrown when a dictionary element type has a converter that returns a type different from the declared type, e.g. when you're deserializing `Dictionary<string, MyClass>`, where

```c#
private interface IClass { }

private class MyClass : IClass { }

private class MyFactory : JsonConverterFactory
{
    public override bool CanConvert(Type typeToConvert)
    {
        return typeToConvert == typeof(IClass) || typeToConvert == typeof(MyClass);
    }

    public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializerOptions options)
    {
        return new MyStuffConverter();
    }
}

private class MyStuffConverter : JsonConverter<IClass>
{
    public override IClass Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return new MyClass();
    }

    public override void Write(Utf8JsonWriter writer, IClass value, JsonSerializerOptions options)
    {
        writer.WriteNumberValue(1);
    }
}
```

In this case, the read method of the converter for `typeof(MyClass)` returns a type `IClass` (not the declared type, `MyClass`) which causes an `InvalidCastException` later in the deserialization flow.

## Customer Impact

The ability to serialize nested concurrent dictionaries is restored: https://github.com/dotnet/runtime/issues/670. 

The deserialization scenario is an edge case, but the fix defends against leaking exceptions were it to occur.

## Regression

No. The serialization change's tests are wide-ranging and covers various permutations of nested dictionaries, so a regression is unlikely.

The deserialization change adds more support without rescinding support. A regression is unlikely.

## Risk

Low, per the regression section above.

### CODE REVIEW

File: src/System.Text.Json/tests/Serialization/DictionaryTests.cs
Comment (ericstj) [inline src/System.Text.Json/tests/Serialization/DictionaryTests.cs:973]: Any comments here or in the test case to summarize what this is doing?  It looks to me like it expands out to a ton of different nesting combinations.  It looks mildly exponential in that each of the 4 outer iterations is more than doubling the size of baseDictionaryTypes.

Comment (layomia) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.cs:973]: I added a comment in the source -

```
// This method generates 316 unique test cases for nested dictionaries up to 4
// levels deep, along with matching JSON, encompassing the various planes of
// dictionaries that can be combined: generic, non-generic, BCL, user-derived,
// immutable, mutable, readonly, concurrent, specialized.
```

Comment (ericstj) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.cs:973]: Are you ok with exponential behavior?  How many cases does it generate now?  What would happen if you change depth?

**Comment by danmoseley** [comment]
Can you link to the change you’re back porting

File: src/System.Text.Json/tests/Serialization/DictionaryTests.cs
Comment (layomia) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.cs:973]: It generates 316 test cases (run as 1 test) now at depth 4, and I don't think the depth should be increased. I think the exponential behavior is okay so long as it remains contained. Would comments discouraging changes to this method that increase the number of tests be okay?

I like this approach because it should hopefully make it easier to catch future changes that break nesting scenarios/permutations that one might not readily consider. If 316 is too much, would this kind of test be better as an outerloop run?

**Comment by layomia** [comment]
https://github.com/dotnet/runtime/pull/784 fixes dotnet/runtime#670 for runtime/master (5.0), but doesn't contain the change on deserialization described [above](https://github.com/dotnet/corefx/pull/42772#issue-352206977). We can consider this PR a back-port + the deserialization change.

File: src/System.Text.Json/tests/Serialization/DictionaryTests.cs
Comment (ericstj) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.cs:973]: Got it.  I’d just comment on the depth to caution on changes.  316 alone isn’t concerning: how much time does it take?

Comment (layomia) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.cs:973]: Done! It takes ~0.35s, taking the total test duration from ~22.5s to ~22.85s.

Comment (ericstj) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.cs:973]: I see, seems reasonable to me.

**Review by ericstj — APPROVED** [review]
LGTM: @ahsonkhan or @steveharter can you also review?

Note that this cannot merge until it is approved for servicing.  Please add the servicing template to the pull request description.

**Comment by danmoseley** [comment]
@ericstj is this ready for tactics consideration now? just noticed it's not tagged but I see the template above

**Comment by ericstj** [comment]
Yeah this is good, @steveharter signed off on same code change in master.

**Comment by jamshedd** [comment]
Approved for March.

**Comment by danmoseley** [comment]
Moved back to 3.1.2 per tactics.

**Pull request merged**
